### DISCUSSION TITLE
Questtion: Whether the BLE device upgrade plug-in is available

### DISCUSSION THREAD

**Discussion by Krcuci** [fwupd/fwupd#7256] [General]
We plan to use fwupd to upgrade the firmware of tws headphone. Do you have the plug-in for BLE device upgrade for reference. Thanks

**Comment by hughsie** [comment]
Yes, we use bluez to communicate with bluetooth and BLE hardware. Can you describe the update protocol e.g. OBEX/GattCharacteristic etc

**Reply by Krcuci** [reply]
Yes, we use the OBEX/GattCharacteristic protocol to communicate. I would also like to ask how to control the upgrade process and content to conform to the upgrade process of hardware equipment. thanks :)

**Reply by hughsie** [reply]
Can you share the Gatt procedure you use now? For instance, how do you get the "current" firmware version, and then switch the device into bootloader mode, and then send packets of the new firmware to it?

**Reply by Krcuci** [reply]
We're still working on the upgrade protocol. Is there any other product of fwupd that supports Ble (non-HID) firmware upgrade at present? If so, we would like to refer to it and keep it consistent.

**Reply by hughsie** [reply]
There's a Qualcomm plugin using BT in the works (@d4s is working on this) but the thing to search for in the code is FuBluezDevice -- there are a lot of helpers there.

**Reply by Krcuci** [reply]
I would like to ask, now I have a tws headset, after connecting the Bluetooth of linux system, how should I operate so that fwupdtool get-devices can obtain the information of this tws headset? Thanks :)

**Reply by d4s** [reply]
@Krcuci please check this example: https://github.com/fwupd/fwupd/tree/wip/d4s/bluez-example

Sorry, cut from the project I'm working with, so might be clumpy.

**Reply by Krcuci** [reply]
I would like to ask the example code you provided, when the computer is connected to the target Bluetooth device, whether the desired device can be obtained by using the command `fwupdtool get-devices`?
